libecole/include/ecole/dynamics/primalsearch.hpp

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <cstddef>
+#include <optional>
+
+#include <nonstd/span.hpp>
+#include <xtensor/xtensor.hpp>
+
+#include "ecole/dynamics/dynamics.hpp"
+
+namespace ecole::dynamics {
+
+/**
+ * An array of variable identifiers in the transformed problem.
+ */
+using VarIds = xt::xtensor<std::size_t, 1>;
+
+/**
+ * A tuple of variable identifiers and variable values.
+ */
+using VarIdsVals = std::tuple<nonstd::span<std::size_t>, nonstd::span<SCIP_Real>>;
+
+class PrimalSearchDynamics : public EnvironmentDynamics<VarIdsVals, std::optional<VarIds>> {
+public:
+	using Action = VarIdsVals;
+	using ActionSet = std::optional<VarIds>;
+
+	PrimalSearchDynamics(int trials_per_node = 1, int depth_freq = 1, int depth_start = 0, int depth_stop = -1);
+
+	std::tuple<bool, ActionSet> reset_dynamics(scip::Model& model) override;
+
+	std::tuple<bool, ActionSet> step_dynamics(scip::Model& model, Action const& action) override;
+
+private:
+	int trials_per_node;
+	int depth_freq;
+	int depth_start;
+	int depth_stop;
+
+	int trials_spent = 0;                 // to keep track of the number of trials during each search
+	SCIP_HEUR* heur = nullptr;            // to tell SCIP where primal solutions come from
+	SCIP_RESULT result = SCIP_DIDNOTRUN;  // the final result of each search (several trials)
```

Comment:
Does `result` need to be an attribute? _i.e._ does it outlive a single `extract` call?

Reply:
It outlives a call to `step_dynamics()`, so I think yes.